### PR TITLE
*: fix 3235

### LINKED ISSUE #3340: `select cast(111 as decimal(1))` is not consistent with MySQL
Please answer these questions before submitting your issue. Thanks!

1. What did you do?
If possible, provide a recipe for reproducing the error.
```
select cast(111 as decimal(1));
```

2. What did you expect to see?
```
mysql> select cast(111 as decimal(2));
+-------------------------+
| cast(111 as decimal(1)) |
+-------------------------+
|                      9 |
+-------------------------+
1 row in set, 1 warning (0.00 sec)
```
3. What did you see instead?
```
mysql> select cast(111 as decimal(1));
+-------------------------+
| cast(111 as decimal(1)) |
+-------------------------+
|                     111 |
+-------------------------+
1 row in set (0.00 sec)
```

4. What version of TiDB are you using (`tidb-server -V`)?
```
Git Commit Hash: 2198d0e7521a5c6c493a4c7af58509b6b277fd6e
UTC Build Time:  2017-05-23 03:07:15
```

### PR DESCRIPTION
**Pull request by XuHuaiyu**
This commit had intented to fix #3340 ,
but it actually fixed the problem described in #3235 .

Now we'll get an error when executing the sql mentioned in #3340 ,
It seems we need a variable in statement context like `OverflowAsWarning` to avoid this.

PTAL @shenli @hanfei1991 @zimulala @tiancaiamao

### CODE REVIEW

File: util/types/datum.go [hunk 1]
```diff
@@ -1048,20 +1048,32 @@ func (d *Datum) convertToMysqlDecimal(sc *variable.StatementContext, target *Fie
...
+	ret.SetValue(dec)
+	return ret, errors.Trace(err)
+}
+
+// produceDecToSpecifiedTp produces a new decimal according to `Tp`.
+func produceDecWithSpecifiedTp(dec *MyDecimal, tp *FieldType, sc *variable.StatementContext) (_ *MyDecimal, err error) {
+	flen, decimal := tp.Flen, tp.Decimal
+	val, err := dec.ToFloat64()
+	if err != nil {
+		dec = nil
```
Comment (tiancaiamao) [inline util/types/datum.go]: `return error.Trace(err)`

**Comment by tiancaiamao** [comment]
PTAL @coocood

File: util/types/datum.go (hunk 1 above)
Comment (XuHuaiyu) [inline util/types/datum.go]: err will not be lost, it will be returned in line 1076

File: util/types/datum.go [hunk 2]
```diff
@@ -1048,20 +1048,32 @@ func (d *Datum) convertToMysqlDecimal(sc *variable.StatementContext, target *Fie
...
-	if target.Flen != UnspecifiedLength && target.Decimal != UnspecifiedLength {
+	dec, err = produceDecWithSpecifiedTp(dec, target, sc)
+	ret.SetValue(dec)
+	return ret, errors.Trace(err)
+}
+
+// produceDecToSpecifiedTp produces a new decimal according to `Tp`.
+func produceDecWithSpecifiedTp(dec *MyDecimal, tp *FieldType, sc *variable.StatementContext) (_ *MyDecimal, err error) {
+	flen, decimal := tp.Flen, tp.Decimal
+	val, err := dec.ToFloat64()
```
Comment (coocood) [inline util/types/datum.go]: Convert decimal to float64 is expensive, we only need to know if the decimal is zero.
Adding an `IsZero` method for `MyDecimal` is better.

Comment (XuHuaiyu) [inline util/types/datum.go]: PTAL @coocood

**Comment by coocood** [comment]
LGTM

**Comment by XuHuaiyu** [comment]
@shenli PTAL

**Comment by tiancaiamao** [comment]
LGTM

**Pull request merged**